src/weights.jl

```diff
@@ -357,8 +357,9 @@ Base.:(==)(x::AbstractWeights, y::AbstractWeights)   = false
 
 Compute the weighted sum of an array `v` with weights `w`, optionally over the dimension `dim`.
 """
-wsum(v::AbstractVector, w::AbstractVector) = dot(v, w)
-wsum(v::AbstractArray, w::AbstractVector) = dot(vec(v), w)
+wsum(v::AbstractVector, w::AbstractVector) = sum(Base.Broadcast.instantiate(Base.Broadcast.Broadcasted(*, (v, w))))
```

Comment:
AFAICT the first argument needs to be `v::AbstractArray`, right? (This will also need tests.)

Reply:
I don't really understand all this `Broadcast....` functions, just copied it from your suggestion. If you say it should work for `AbstractArray` as well, I can modify the signature.
As for tests, again not sure what's the proper test should be. A simple test with weighted sum of vector-of-vectors is already in this PR.

Reply:
I was just thinking about a test that passes a matrix as first argument. The existing methods accepts them, so that should continue to be supported after this PR.

Reply:
Tried changing to `AbstractArray`, but this method fails for multidimensional arrays anyway: with `"arrays could not be broadcast to a common size; got a dimension with lengths 2 and 4`.

Reply:
Have you tried with `vec(v)` as in the original method?